wikitools/code_block_parser.py

```diff
@@ -0,0 +1,117 @@
+import typing
+
+
+class CodeTag(typing.NamedTuple):
+    start: int
+    len: int
+
+
+class CodeBlock(typing.NamedTuple):
+    """
+    A Markdown code block (monospace font + gray backround). Could be one of the following:
+        - Inline: `text` or `` text `` (the latter is used for snippets containing `)
+        - Multiline: ``` followed by anything else until the next ``` block
+
+    Examples:
+
+        ```test test test```
+        ^ start            ^ end
+
+        `this is a very long phrase`
+        ^ start                    ^ end
+
+    Similarly to the HTML comments, links in code blocks are not accounted for.
+    """
+
+    start: int  # 0-based position of the first character of the block's opening tag
+    end: int  # 0-based position of the last character of the block's closing tag
+    tag_len: int
+
+    @property
+    def is_multiline(self):
+        return self.end == -1 and self.tag_len == 3
+
+    def contains(self, other: 'CodeBlock'):
+        return (
+            self.start < other.start and
+            self.end > other.end and
+            self.tag_len > other.tag_len
+        )
+
+
+class CodeBlockParser:
+    def __init__(self):
+        self.__in_multiline = False
+
+    @property
+    def in_multiline(self) -> bool:
+        return self.__in_multiline
+
+    def parse(self, line: str) -> typing.List[CodeBlock]:
+        blocks: typing.List[CodeBlock] = []
+
+        tag_stack = []
+        if self.__in_multiline:
+            tag_stack.append(CodeTag(start=-1, len=3))
+
+        i = 0
+        while i < len(line):
+            if line[i] != '`':
+                i += 1
+                continue
+
+            cnt = 0
+            while i + cnt < len(line) and line[i + cnt] == '`':
+                cnt += 1
+
+            # just closed the code block
+            if tag_stack and tag_stack[-1].len == cnt:
+                opening_tag = tag_stack.pop()
```

Comment:
line 69 is under `while True`, line 78 isn't. to put it another way, the function works as follows:

- while possible, extract a tag (any sequence of grave accents) greedily.
- if the previous tag we've found is of the same length, this means that we've just found a closing tag, and we can say that we've found a code block that spans from `the start of the previous tag (tag_stack[-1])` to `the end of the currently parsed tag`. right here, we remove the opening tag from the stack, closing the code block.
- if not, we just push the newfound tag onto the stack and continue.

when there isn't anything else on the line, look at the **tags** which we couldn't close. there could be two cases:

1. there is none = everything was properly closed, and we need to consider which code block takes priority over which (that's line 89)
2. there is something, meaning there wasn't a pair for one or more tags on the line (that's line 77)

for 2), I consider two options:

- the line starts with a multiline code block which doesn't close on this line, and we don't really care what's inside of it, saying "it's multiline anyway"
- some other sequence of tags. if there are no completed code blocks, it could be anything, so let's just return what we have successfully parsed

Reply:
actually the part where we deal with non-empty stack needs to be fixed

Reply:
as for the tests, I guess they are simple and there's only one non-closed code tag at time, so `len(tag_stack) == 1`. added one to show it's not always the case

Reply:
[`38135ff` (#6)](https://github.com/Walavouchey/osu-wiki-tools/pull/6/commits/38135ffbd44f47546d5ee651d40406a229328ec4) and [`b98e46b` (#6)](https://github.com/Walavouchey/osu-wiki-tools/pull/6/commits/b98e46b9e9511828e7444c11b254dc97bc789980) address this